### PR TITLE
Yain's Completed IC

### PR DESCRIPTION
**Pull request by yainporter**
I think it is interesting that I still get tripped up on certain MVC principles. I am working on differentiating between MVC infractions vs best practices vs coder preference. For example:

This link [here](https://backend.turing.edu/module2/lessons/intro_to_mvc#:~:text=Logic%20%E2%80%9CResponsibilities%E2%80%9D) says that we should not call class methods in the Views, however in the MVC class with Abdul, he created a class method in the Special class to call on @model.specials.class_method. Is that an MVC infraction?

I probably should have done more testing for edge cases, sad path, etc, but I kind of just wanted to get the IC done, so hopefully this is okay.

### CODE REVIEW

**Comment by juliet-e** [comment]
> I think it is interesting that I still get tripped up on certain MVC principles. I am working on differentiating between MVC infractions vs best practices vs coder preference. For example:
> 
> This link [here](https://backend.turing.edu/module2/lessons/intro_to_mvc#:~:text=Logic%20%E2%80%9CResponsibilities%E2%80%9D) says that we should not call class methods in the Views, however in the MVC class with Abdul, he created a class method in the Special class to call on @model.specials.class_method. Is that an MVC infraction?
> 
> I probably should have done more testing for edge cases, sad path, etc, but I kind of just wanted to get the IC done, so hopefully this is okay.

Great question about class methods! It's definitely a controversial question and gets into the weeds of MVC/OOP technicalities, so we don't spend too much time on it because it's not as important as other MVC principles (i.e. encapsulate data querying in the model). BUT, the way I think about it is:
1. It's okay to call a class method on a collection in the view (i.e. `object.specials.class_method`)
2. If we're calling a class method on a Class name itself, it's typically better to instead call that from the controller and make an instance variable, because views should be responsible for just handling instances (sent from the controller). From the perspective of strict Rails OOP, the view should be just taking these objects and calling methods on them. Calling a class method departs from this practice and the view technically shouldn't "know" about classes -- it just handles objects. But it's okay to call a class method on a collection of instances like is done in my first example. 

Out in the world, you could absolutely see a class method called from the view on the class name itself, and I personally wouldn't consider that a true MVC violation but I'm sure others would disagree with me!

File: app/controllers/contestant_projects_controller.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class ContestantProjectsController < ApplicationController
+  def create
+    contestant_project = ContestantProject.new(contestant_project_params)
+    if contestant_project.save
```
Comment (juliet-e) [inline app/controllers/contestant_projects_controller.rb:4]: Nice additional error handling! We don't require this on ICs but it's great to practice when you have the time!

File: app/models/project.rb [hunk 2]
```diff
@@ -1,3 +1,13 @@
 class Project < ApplicationRecord
   belongs_to :challenge
+  has_many :contestant_projects
+  has_many :contestants, through: :contestant_projects
+
+  def number_of_contestants
```
Comment (juliet-e) [inline app/models/project.rb:6]: Excellent encapsulation

File: app/views/contestants/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<h1>All Contestants:</h1>
+<ul>
+  <% @contestants.each do |contestant| %>
+    <div id="contestant-<%= contestant.id %>">
+      <li><%= contestant.name %></li>
+      <li>Projects: <%= contestant.names_of_projects.join(", ") %></li>
```
Comment (juliet-e) [inline app/views/contestants/index.html.erb:6]: Nice work making the comma separated list! And as it is display logic, it's great for the view

File: config/routes.rb [hunk 4]
```diff
@@ -3,4 +3,11 @@
 
   # Defines the root path route ("/")
   # root "articles#index"
+
+  get "projects/:id", to: "projects#show", as: :show_projects
+
+  get "contestants/", to: "contestants#index", as: :contestants
+
+  post "contestant_projects/", to: "contestant_projects#create", as: :create_contestant_projects
```
Comment (juliet-e) [inline config/routes.rb:11]: Great RESTful routing! Just a reminder that you can include IDs in the path and that's still RESTful as well. If you didn't want to have to add a hidden field in your form for project ID, you could use this path: `"/projects/:project_id/contestants"` -- either are RESTful

File: spec/features/contestants/index_spec.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      expect(page).to have_content("All Contestants:")
+      expect(page).to have_content("Gretchen Jones")
+      expect(page).to have_content("Erin Robertson")
+
+      within "#contestant-#{@jay.id}" do
+        expect(page).to have_content("Jay McCarroll")
+        expect(page).to have_content("Projects: Litfit")
+      end
+
+      within "#contestant-#{@kentaro.id}" do
```
Comment (juliet-e) [inline spec/features/contestants/index_spec.rb:40]: Excellent use of within blocks to make these assertions meaningful

File: spec/models/project_spec.rb [hunk 6]
```diff
@@ -7,4 +7,43 @@
...
+      it "returns the number of contestants for a Project" do
+        expect(@news_chic.number_of_contestants).to eq(2)
+        expect(@boardfit.number_of_contestants).to eq(3)
+      end
+    end
+
+    describe "#avg_experience" do
+      it "returns the average experience of Contestants" do
+        expect(@news_chic.avg_experience).to eq(12.5)
+        expect(@news_chic.avg_experience).to eq(12.5)
```
Comment (juliet-e) [inline spec/models/project_spec.rb:45]: Well done testing at the unit and integration level!

File: config/routes.rb (hunk 4 above)
Comment (yainporter) [reply config/routes.rb:11]: I think I remember asking a question on this, the action would be contestants#create, as long as there is not already a post "contestants/", to: "contestants#create", right?

Comment (juliet-e) [reply config/routes.rb:11]: No, the action looks great and can stay the same! And to be clear, the path is also okay. I'm just sharing an alternative in case you want a path parameter in there to avoid having to pass a hidden field in the form
